File: Sagittarius.jl/src/cluster.jl
module Cluster

using Distributed
using ..Program
using ..Physics
using ..Solver
using ..StructuredLogging: log_event

export setup_workers, run_parallel_sweep

function setup_workers(n::Int)
    log_event("cluster_setup_start"; n_workers=n)
    if nprocs() < n
        addprocs(n - nprocs())
    end
    # Sagittarius loading must be done from Main or via proper package load
    worker_count = nprocs()
    log_event("cluster_setup_finish"; n_workers=worker_count)
    return worker_count
end

"""
    run_parallel_sweep(sim_configs, psi0, t_span, observables)

Runs a list of simulation configurations in parallel.
`sim_configs` should be a list of dicts or objects that can be 
used to construct/run simulations on workers.
"""
function run_parallel_sweep(run_func, params)
    # run_func is a function that takes one element of params
    # and returns the result.
    return pmap(run_func, params)
end

end # module Cluster

File: sagittarius_py/sagittarius/cluster.py
from typing import List, Callable, Any, Dict
import numpy as np
from .api import Simulation, SimulationResult
from .runtime import get_julia, log_event

class ParallelSimulation:
    """
    Handles parallel parameter sweeps by offloading the map-reduce 
    to Julia's Distributed module.
    """
    
    def __init__(self, n_workers: int = 4):
        self.n_workers = n_workers
        # Initialize workers and load Sagittarius only when cluster execution is requested.
        jl, sgr = get_julia()
        log_event("cluster_setup_start", n_workers=n_workers)
        sgr.Cluster.setup_workers(n_workers)
        log_event("cluster_setup_finish", n_workers=n_workers)
        
    def map(self, sim_func: str, params: List[Any]) -> List[Any]:
        """
        Runs a parallel sweep using a Julia function name and a list of parameters.
        
        sim_func: Name of a Julia function (already defined on workers) 
                  that takes one parameter.
        params: List of parameters (must be serializable to Julia).
        """
        # We call pmap via Distributed module
        jl, _ = get_julia()
        results = list(jl.Distributed.pmap(jl.seval(sim_func), params))
        return results

    def sweep_omega(self, register, base_sequence, omegas: List[float], t_end: float) -> List[SimulationResult]:
        """
        A high-level helper to sweep Rabi frequency.
        This demonstrates how to use the clustered solver for a common task.
        """
        # For a real implementation, we would broadcast the register and sequence once,
        # then pmap only the variable parameter.
        
        # Prototype: run sequentially in Python for now to define the interface,
        # but mark for Phase 4.2 full Julia-side offloading.
        results = []
        for w in omegas:
            # Update omega in a copy of the sequence
            # ... construction logic ...
            pass
            
        raise NotImplementedError("Full Julia-side sweep offloading is in development for Phase 4.2")
